nixos/modules/services/web-apps/audiobookshelf.nix

```diff
@@ -0,0 +1,113 @@
+{ pkgs
+, lib
+, config
+, ...
+}:
+
+with lib; let
+  cfg = config.services.audiobookshelf;
+in
+
+{
+  options.services.audiobookshelf = {
+    enable = mkEnableOption (lib.mdDoc  "Audiobookshelf, self-hosted audiobook and podcast server");
+
+    host = mkOption {
+      type = types.str;
+      default = "127.0.0.1";
+      description = lib.mdDoc ''
+        The host address that Audiobookshelf will listen on.
+      '';
+    };
+
+    port = mkOption {
+      type = types.port;
+      default = 8000;
+      description = lib.mdDoc ''
+        The port that Audiobookshelf will listen on.
+      '';
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = "audiobookshelf";
+      description = lib.mdDoc ''
+        User account under which Audiobookshelf runs.
+      '';
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "audiobookshelf";
+      description = lib.mdDoc ''
+        Group under which Audiobookshelf runs.
+      '';
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        Whether to open the firewall for the port in {option}`services.audiobookshelf.port`.
+      '';
+    };
+
+    configDir = mkOption {
+      type = types.str;
+      default = "/etc/audiobookshelf";
+      description = lib.mdDoc ''
+        Configuration directory Audiobookshelf will use.
+      '';
+    };
+
+    metadataDir = mkOption {
+      type = types.str;
+      default = "/var/lib/audiobookshelf";
+      description = lib.mdDoc ''
+        Metadata directory Audiobookshelf will use.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    networking.firewall.allowedTCPPorts = mkIf cfg.openFirewall [ cfg.port ];
+
+    users.groups = mkIf (cfg.group == "audiobookshelf") {
+      audiobookshelf = { };
+    };
+
+    users.users = mkIf (cfg.user == "audiobookshelf") {
+      audiobookshelf = {
+        group = cfg.group;
+        description = "Audiobookshelf Daemon user";
+        isSystemUser = true;
+      };
+    };
+
+    systemd.services.audiobookshelf = {
+      description = "Audiobookshelf is a self-hosted audiobook and podcast server";
+
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ];
+      after = [ "network-online.target" ];
+
+      serviceConfig = {
```

Comment:
I've just made my own little service config for this before I realised this PR was already here, and if I may offer a suggestion, it might be better to use systemd to pass the environment variables directly instead of using the bash script for the service for the sake of efficiency and reducing dependencies:

```
systemd.services.audiobookshelf = {
  enable = true;
  description = "Self-hosted audiobook server for managing and playing audiobooks";
  environment = {
    NODE_ENV = "production";
    SOURCE = "nixpkgs";
    FFMPEG_PATH = "${pkgs.ffmpeg-full}/bin/ffmpeg";
    FFPROBE_PATH = "${pkgs.ffmpeg-full}/bin/ffprobe";
    TONE_PATH = "${pkgs.tone}/bin/tone";
    CONFIG_PATH="${cfg.configDir}";
    METADATA_PATH="${cfg.metadataDir}";
    PORT="${toString cfg.port}";
    HOST="${cfg.host}";
  };
  serviceConfig = {
    Type = "simple";
    WorkingDirectory = "/var/lib/audiobookshelf";
    ExecStart = "${pkgs.nodejs}/bin/node ${pkgs.audiobookshelf}/opt/index.js";
    ExecReload = "${pkgs.util-linux}/bin/kill -HUP $MAINPID";
    Restart = "always";
    User = "audiobookshelf";
    Group = "audiobookshelf";
  };
  wantedBy = [ "multi-user.target" ];
  requires = [ "network.target" ];
};
```
This eliminates the need for invoking `bash` in the service file. While I realise the advantage of the wrapper for the regular package, I think the service file would be more efficient without the use of bash.

Reply:
@savyajha  I agree with most of your suggestions, but disagree on this one. I don't think the shell script in `wrapper.nix` adds any real overhead, and I think that it's more difficult for maintenance if these details are defined in both `wrapper.nix` and this module.

I am pro simplicity. My personal service that I recently created looks like this:
```nix
systemd.services.audiobookshelf = {
  description = "Audiobookshelf is a self-hosted audiobook and podcast server";
  after = [ "network.target" ];
  wantedBy = [ "multi-user.target" ];

  serviceConfig = {
    Type = "simple";
    User = cfg.user;
    Group = cfg.group;
    WorkingDirectory = cfg.dataDir;
    ExecStart = "${cfg.package}/bin/audiobookshelf --host ${cfg.host} --port ${toString cfg.port}";
    Restart = "on-failure";
  };
};
```

Note that I do not manually define the config and metadata directories, and I instead use the more standard `dataDir`. As defined in [`wrapper.nix`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/audiobookshelf/wrapper.nix), config and metadata directories default to `$PWD/config` and `$PWD/metadata` respectively. In practice, this will be `/var/lib/audiobookshelf/{config,metadata}`.

Reply:
@wietsedv Fair enough. I like using systemd wherever possible over bash, but that's merely personal preference. However, it would be good to use the tmpfiles snippet I mentioned in another comment if you're not going to use `StateDirectory` so that the folder is created at first run.